src/io/flutter/run/test/DebugTestRunner.java

```diff
@@ -0,0 +1,195 @@
+import io.flutter.run.PositionMapper;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Runs a Flutter test configuration in the debugger.
+ */
+public class DebugTestRunner extends GenericProgramRunner {
+
+  @NotNull
+  @Override
+  public String getRunnerId() {
+    return "FlutterDebugTestRunner";
+  }
+
+  @Override
+  public boolean canRun(@NotNull String executorId, @NotNull RunProfile profile) {
+    if (!DefaultDebugExecutor.EXECUTOR_ID.equals(executorId) || !(profile instanceof TestConfig)) {
+      return false;
+    }
+    final TestConfig config = (TestConfig) profile;
+    return config.getFields().getScope() == TestFields.Scope.FILE;
+  }
+
+  @Nullable
+  @Override
+  protected RunContentDescriptor doExecute(@NotNull RunProfileState state, @NotNull ExecutionEnvironment env)
+    throws ExecutionException {
+    return runInDebugger((TestLaunchState) state, env);
+  }
+
+  protected RunContentDescriptor runInDebugger(@NotNull TestLaunchState launcher, @NotNull ExecutionEnvironment env)
+    throws ExecutionException {
+
+    // Start process and create console.
+    final ExecutionResult executionResult = launcher.execute(env.getExecutor(), this);
+    final ObservatoryConnector connector = new Connector(executionResult.getProcessHandler());
+
+    // Set up source file mapping.
+    final DartUrlResolver resolver = DartUrlResolver.getInstance(env.getProject(), launcher.getTestFileOrDir());
+    final PositionMapper.Analyzer analyzer = PositionMapper.Analyzer.create(env.getProject(), launcher.getTestFileOrDir());
+    final PositionMapper mapper = new PositionMapper(env.getProject(), launcher.getPubRoot().getRoot(), resolver, analyzer);
+
+    // Create the debug session.
+    final XDebuggerManager manager = XDebuggerManager.getInstance(env.getProject());
+    final XDebugSession session = manager.startSession(env, new XDebugProcessStarter() {
+      @Override
+      @NotNull
+      public XDebugProcess start(@NotNull final XDebugSession session) {
+        return new TestDebugProcess(session, executionResult, resolver, connector, mapper);
+      }
+    });
+
+    return session.getRunContentDescriptor();
+  }
+
+  /**
+   * Provides observatory URI, as received from the test process.
+   */
+  private static final class Connector implements ObservatoryConnector {
+    private final ProcessListener listener;
+    private String observatoryUri;
+
+    public Connector(ProcessHandler handler) {
+      listener = new ProcessAdapter() {
+        @Override
+        public void onTextAvailable(ProcessEvent event, Key outputType) {
+          if (!outputType.equals(ProcessOutputTypes.STDOUT)) {
+            return;
+          }
+
+          final String text = event.getText().trim();
+          if (FlutterInitializer.isVerboseLogging()) {
+            LOG.info("[<-- " + text + "]");
+          }
+
+          if (!text.startsWith("[{") || !text.endsWith("}]")) {
```

Comment:
Yes, I used the same JSON format as daemon mode. But currently there's only one message, and it's only printed when using --machine and --start-paused.